test/extended/networking/commatrix.go

```diff
@@ -0,0 +1,215 @@
+
+		workerNFT, err := workerMat.ToNFTables()
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("apply firewall on master nodes")
+		err = applyRulesToNode(cs, "master", masterNFT)
+		o.Expect(err).ToNot(o.HaveOccurred())
+
+		if !isSNO {
+			g.By("apply firewall on worker nodes")
+			err = applyRulesToNode(cs, "worker", workerNFT)
+			o.Expect(err).ToNot(o.HaveOccurred())
+		}
+	})
+})
+
+func applyRulesToNode(cs *client.ClientSet, role string, NFTtable []byte) error {
+	utilsHelpers := utils.New(cs)
+	ns := consts.DefaultDebugNamespace + "firewall"
+	err := utilsHelpers.CreateNamespace(ns)
+	if err != nil {
+		return err
+	}
+
+	defer func() error {
+		err := utilsHelpers.DeleteNamespace(ns)
+		if err != nil {
+			return err
+		}
+		return nil
+	}()
+
+	nodesList, err := cs.CoreV1Interface.Nodes().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to list nodes: %w", err)
+	}
+
+	artifactsDir := filepath.Join(exutil.ArtifactDirPath(), "commatrix")
+	err = os.MkdirAll(artifactsDir, 0755)
+	if err != nil {
+		return fmt.Errorf("failed to create artifactsDir %w", err)
+	}
+
+	for _, node := range nodesList.Items {
+		if types.GetNodeRoles(&node)[0] != role {
+			continue
+		}
+
+		debugPod, err := utilsHelpers.CreatePodOnNode(node.Name, ns, consts.DefaultDebugPodImage)
+		if err != nil {
+			return fmt.Errorf("failed to create debug pod on node %s: %w", node.Name, err)
+		}
+
+		defer func() {
+			err := utilsHelpers.DeletePod(debugPod)
+			if err != nil {
+				fmt.Printf("failed cleaning debug pod %s: %v", debugPod, err)
+			}
+		}()
+
+		command := []string{
+			"bash", "-c",
+			fmt.Sprintf("mkdir -p nft && echo '%s' > nft/firewall.nft", string(NFTtable)),
+		}
+		_, err = utilsHelpers.RunCommandOnPod(debugPod, command)
+		if err != nil {
+			return fmt.Errorf("failed to save rule set to file on node %s: %w", node.Name, err)
+		}
+
+		installCmd := []string{"bash", "-c", "yum update -y && yum install -y nftables"}
+		_, err = utilsHelpers.RunCommandOnPod(debugPod, installCmd)
+		if err != nil {
+			return fmt.Errorf("failed to save rule set to file on node %s: %w", node.Name, err)
+		}
+
+		command = []string{
+			"bash", "-c", "/usr/sbin/nft -f nft/firewall.nft",
+		}
+		_, err = utilsHelpers.RunCommandOnPod(debugPod, command)
+		if err != nil {
+			return fmt.Errorf("failed to apply rule set %q on node %s: %w", command, node.Name, err)
+		}
+
+		// Check the output of nft list ruleset
+		command = []string{"bash", "-c", "/usr/sbin/nft list ruleset"}
+		output, err := utilsHelpers.RunCommandOnPod(debugPod, command)
+		if err != nil {
+			return fmt.Errorf("failed to list NFT ruleset on node %s: %w", node.Name, err)
+		}
+		fmt.Printf("Command output: %s\n", string(output))
+
+		if len(output) == 0 {
+			return fmt.Errorf("no output from 'nft list ruleset' on node %s: ", node.Name)
+		}
+		command = []string{
+			"bash", "-c",
+			"/usr/sbin/nft list ruleset > /etc/nftables.conf",
```

Comment:
Is there a reason to list the ruleset twice?

Reply:
one for the output and the other for the save after reboot

Reply:
i did twice since the output is so long somtimes and i cant echo it to file so better to do like this